optimum/habana/transformers/models/chatglm/modeling_chatglm.py

```diff
@@ -0,0 +1,1862 @@
+"""PyTorch ChatGLM model."""
```

Comment:
@mengker33 , 
make style is failing on this file. You'll also need to add the "license" on all new files that you created.
Here is a suggested change.
diff --git a/optimum/habana/transformers/models/chatglm/modeling_chatglm.py b/optimum/habana/transformers/models/chatglm/modeling_chatglm.py
index 8e799706..ee9a2414 100644
--- a/optimum/habana/transformers/models/chatglm/modeling_chatglm.py
+++ b/optimum/habana/transformers/models/chatglm/modeling_chatglm.py
@@ -1,12 +1,20 @@
-"""PyTorch ChatGLM model."""
-
-"""
-Adapted from the following sources:
-https://huggingface.co/THUDM/chatglm2-6b/blob/main/modeling_chatglm.py
-https://huggingface.co/THUDM/chatglm3-6b/blob/main/modeling_chatglm.py
-https://huggingface.co/THUDM/glm-4-9b-chat/blob/main/modeling_chatglm.py
-"""
-
+# coding=utf-8
+# Copyright 2023 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+###############################################################################
+# Copyright (C) 2022-2024 Habana Labs, Ltd. an Intel Company
+###############################################################################
 import copy
 import json
 import math
@@ -36,6 +44,14 @@ from ...modeling_attn_mask_utils import _gaudi_prepare_4d_causal_attention_mask
 from .configuration_chatglm import ChatGLMConfig


+"""PyTorch ChatGLM model."""
+
+"""
+Adapted from the following sources:
+https://huggingface.co/THUDM/chatglm2-6b/blob/main/modeling_chatglm.py
+https://huggingface.co/THUDM/chatglm3-6b/blob/main/modeling_chatglm.py
+https://huggingface.co/THUDM/glm-4-9b-chat/blob/main/modeling_chatglm.py
+"""
 try:
     from habana_frameworks.torch.hpex.kernels import FusedSDPA
 except ImportError:

Reply:
Thanks for the suggestion! I found the cause for "make style" failure is that """PyTorch ChatGLM model.""" should be followed by import part instead of """ Adapted from the following sources: xxx""".
As for the license, since chatglm is a customized model and there is no license directly shown in their modeling codes, (see https://huggingface.co/THUDM/chatglm2-6b/blob/main/modeling_chatglm.py
https://huggingface.co/THUDM/chatglm3-6b/blob/main/modeling_chatglm.py
https://huggingface.co/THUDM/glm-4-9b-chat/blob/main/modeling_chatglm.py)
I am not sure if it fine to add license at the beginning of modeling python file, if you think it ok, I will add it later.

Reply:
@regisss , The original files don't have a license or copyright. Please advise if we need to add a license or copyright.

@mengker33 , I think some of the code is take from modeling _lama*.py ? If so, please add that to the text.

Reply:
The modeling code of `chatglm2-6b` and `chatglm3-6b` is under Apache-2.0:
- https://huggingface.co/THUDM/chatglm3-6b#%E5%8D%8F%E8%AE%AE-license
- https://huggingface.co/THUDM/chatglm2-6b#%E5%8D%8F%E8%AE%AE
So I think you can add the license for these ones.

For GLM4, the license is different so I'm not sure we can do that: https://huggingface.co/THUDM/glm-4-9b-chat#%E5%8D%8F%E8%AE%AE

Reply:
> # coding=utf-8
> +# Copyright 2023 The HuggingFace Team. All rights reserved.
> +#
> +# Licensed under the Apache License, Version 2.0 (the "License");
> +# you may not use this file except in compliance with the License.
> +# You may obtain a copy of the License at
> +#
> +# http://www.apache.org/licenses/LICENSE-2.0
> +#
> +# Unless required by applicable law or agreed to in writing, software
> +# distributed under the License is distributed on an "AS IS" BASIS,
> +# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> +# See the License for the specific language governing permissions and
> +# limitations under the License.
> +###############################################################################
> +# Copyright (C) 2022-2024 Habana Labs, Ltd. an Intel Company
> +###############################################################################

Only gaudi_chatglm_repeat_kv is copied from optimum-habana's llama optimization, and it has been explained in the comment of this function.

Reply:
> The modeling code of `chatglm2-6b` and `chatglm3-6b` is under Apache-2.0:
> 
> * https://huggingface.co/THUDM/chatglm3-6b#%E5%8D%8F%E8%AE%AE-license
> * https://huggingface.co/THUDM/chatglm2-6b#%E5%8D%8F%E8%AE%AE
>   So I think you can add the license for these ones.
> 
> For GLM4, the license is different so I'm not sure we can do that: https://huggingface.co/THUDM/glm-4-9b-chat#%E5%8D%8F%E8%AE%AE

GLM4-9b series model was merged and standardized in Hugging Face transformers last week,  I may need to refine the codes accordingly.